lib/router/src/utils.ts

```diff
@@ -37,27 +38,48 @@ interface Args {
   [key: string]: any;
 }
 
-const deepDiff = (value: any, update: any): any => {
-  if (deepEqual(value, update)) return undefined;
+export const DEEPLY_EQUAL = Symbol('Deeply equal');
+export const deepDiff = (value: unknown, update: unknown): any => {
   if (typeof value !== typeof update) return update;
-  if (Array.isArray(value)) return update;
-  if (update && typeof update === 'object') {
-    return Object.keys(update).reduce((acc, key) => {
-      const diff = deepDiff(value[key], update[key]);
-      return diff === undefined ? acc : Object.assign(acc, { [key]: diff });
+  if (deepEqual(value, update)) return DEEPLY_EQUAL;
+  if (Array.isArray(value) && Array.isArray(update)) {
+    return value.length > update.length
+      ? [...new Array(value.length)].map((_, i) => update[i])
```

Comment:
I think the spread into a new array is unnecessary here and creating two new arrays instead of one.

Also, I realize the implicit behavior of accessing an out-of-bounds index on an array returned `undefined`, but I think something like this could be a bit more explicit:
`[...update, ...Array(value.length - update.length).fill(undefined)]`

Reply:
I like it. We don't even need the `fill` because undefined is what you get when spreading a sparse array.